homeassistant/components/apcupsd/coordinator.py

```diff
@@ -73,11 +73,16 @@ def __init__(self, hass: HomeAssistant, host: str, port: int) -> None:
         self._host = host
         self._port = port
 
+    @property
+    def unique_device_id(self) -> str:
+        """Return a unique ID of the device, which is the serial number (if available) or the config entry ID."""
+        return self.data.serial_no or self.config_entry.entry_id
```

Comment:
I think you can set this in the constructor to an attribute itself so it doesn't have to evaluate for every entity again.

Reply:
I don't think so. Judging from the source code:

```c
   if (ups->UPS_Cap[CI_SERNO])
      s_write(ups, "SERIALNO : %s\n", ups->serial);
```

(where `UPS_Cap` is the static capability array of the UPS).

Updated here. PTAL.

Reply:
> do an initial refresh in the constructor, but I think we generally rely on await coordinator.async_config_entry_first_refresh() in the async_setup_entry?

Yes, what does the code currently do?